main.go

```diff
@@ -136,9 +136,67 @@ func main() {
 		prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{}),
 		prometheus.NewGoCollector(),
 	)
-	go telemetryServer(ksmMetricsRegistry, opts.TelemetryHost, opts.TelemetryPort)
 
-	serveMetrics(ctx, kubeClient, storeBuilder, opts, opts.Host, opts.Port, opts.EnableGZIPEncoding)
+	var g run.Group
+
+	m := metricshandler.New(
+		opts,
+		kubeClient,
+		storeBuilder,
+		opts.EnableGZIPEncoding,
+	)
+	// Run MetricsHandler
+	{
+		ctxMetricsHandler, cancel := context.WithCancel(ctx)
+		g.Add(func() error {
+			return m.Run(ctxMetricsHandler)
+		}, func(error) {
+			cancel()
+		})
+	}
+
+	telemetryMux := buildTelemetryServer(ksmMetricsRegistry)
+	telemetryServer := http.Server{Handler: telemetryMux}
+	telemetryListenAddress := net.JoinHostPort(opts.TelemetryHost, strconv.Itoa(opts.TelemetryPort))
+	telemetryLn, err := net.Listen("tcp", telemetryListenAddress)
+	if err != nil {
+		klog.Fatalf("Failed to create Telemetry Listener: %v", err)
+	}
+	metricsMux := buildMetricsServer(kubeClient, storeBuilder, m, opts)
+	metricsServer := http.Server{Handler: metricsMux}
+	metricsServerListenAddress := net.JoinHostPort(opts.Host, strconv.Itoa(opts.Port))
+	metricsServerLn, err := net.Listen("tcp", metricsServerListenAddress)
+	if err != nil {
+		klog.Fatalf("Failed to create MetricsServer Listener: %v", err)
+	}
+
+	// Run Telemetry server
+	{
+		g.Add(func() error {
+			klog.Infof("Starting kube-state-metrics self metrics server: %s", telemetryListenAddress)
+			return telemetryServer.Serve(telemetryLn)
+		}, func(error) {
+			ctxShutDown, cancel := context.WithTimeout(ctx, 3*time.Second)
+			defer cancel()
+			telemetryServer.Shutdown(ctxShutDown)
```

Comment:
Does this also close the net.Listener? If not then we need to explicitly close that one as well.

Reply:
Also linked to the previous comment:

> Shutdown gracefully shuts down the server without interrupting any active connections. Shutdown works by first closing all open listeners, then closing all idle connections, and then waiting indefinitely for connections to return to idle and then shut down. If the provided context expires before the shutdown is complete, Shutdown returns the context's error, otherwise it returns any error returned from closing the Server's underlying Listener(s). 

As I understand it, it closes the Listener first. The `waiting indefinitely` part is the reason I am using `.WithTimeout`.

Reply:
got it :+1: